### PR TITLE
Add React tester

### LINKED ISSUE #121: Automatic tests for each supported React version
## Are you reporting a feature request or a bug?

Task

## Provide detailed reproduction steps (if any)

We have concluded with @Comandeer that it will be worth having integration tests which check if CKEditor 4 works correctly with all supported React versions - `major`s only I suppose. Generally, it should work be we don't check it too often. This will:

* Protect us from any backwards incompatibilities when introducing changes.
* Allow us quickly adding new tests whenever new React version is out.

It could be done probably in the form of simple application for each React version which will use our component (there are few ways to create such app - manual, CRA, etc. but we should get with the most common one). Also it should be fairly easy to add new tests for new versions.

---

If it works fine we should do the same for other integrations.

### PR DESCRIPTION
**Pull request by Comandeer**
I've added a simple script that fetches available versions of React, find only ones that satisfies semver range defined in `peerDependencies` of the package and then perform tests on every compatible version (by copying the package to temporary dir).

Closes #121.

### CODE REVIEW

File: package.json
Comment (f1ames) [inline package.json:18]: I guess we could improve naming here, one may wonder what is the difference between `test` and `test:react`. Maybe something like `test:versions`, `test:multiple` (both doesn't sound good enough still but maybe you will be able to come up wit better naming)?

File: scripts/react-tester.js
Comment (f1ames) [inline scripts/react-tester.js:44]: Is it some React issue or our integration? If it's our integration we should report an issue. If React maybe some external reference could be useful?

File: scripts/react-tester.js
Comment (f1ames) [inline scripts/react-tester.js:17]: Yes it is! :smile: Or more like "Ultimate React CKEditor 4 Integration Tester" :stuck\_out\_tongue\_winking\_eye:

**Review by f1ames — CHANGES_REQUESTED** [review]
Looks good :+1: 

I have two suggestions to improve the usability. It now tests all versions (starting from `16.0.0`) so major, minor and patches which is a lot and on CI takes more than 40 minutes which makes it kind of unusable. For sure we should test each major (latest one, so `16.14.0` and `17.0.1` ATM), and maybe each minor (again latest, so `16.1.1`, `16.2.0`, `16.3.2`, etc). Still it will be around 15 version for now. But testing each patch version doesn't make sense IMHO. The test run should not take more than 5 - 10 minutes so we get fast response after pushing PRs.

The other thing is that it uses dev integration version (I know that's how tests are run now via `test:browser` command), but I think it will be more reasonable to test production version/build. WDYT?

**Comment by Dumluregn** [comment]
Rebased onto latest `master`.

File: scripts/react-tester.js
Comment (Dumluregn) [reply scripts/react-tester.js:44]: It's React - added reference to an issue: https://github.com/facebook/react/issues/14208.

**Comment by Dumluregn** [comment]
> It now tests all versions (starting from 16.0.0) so major, minor and patches which is a lot and on CI takes more than 40 minutes which makes it kind of unusable. For sure we should test each major (latest one, so 16.14.0 and 17.0.1 ATM), and maybe each minor (again latest, so 16.1.1, 16.2.0, 16.3.2, etc). Still it will be around 15 version for now. But testing each patch version doesn't make sense IMHO. The test run should not take more than 5 - 10 minutes so we get fast response after pushing PRs.

I've limited versions tested according to this suggestion, so only the latest patch version for each minor release is tested. It cuts the number of builds from 33 to 15. Not sure if we need e.g. config option to enable also full run?

> The other thing is that it uses dev integration version (I know that's how tests are run now via test:browser command), but I think it will be more reasonable to test production version/build. WDYT?

I think that's a very reasonable idea and also that we should add it as an option for the "normal" test run - we could then have two runs on bender (built version and dev version like for CKE4), so I'd rather extract it so a separate issue (this would also a good idea in other repos), WDYT?

BTW I've encountered some issues during rebasing this PR, so I had to downgrade back to npm 6 for now (that's why there are so many removed lines from `package-lock.json`).

**Comment by f1ames** [comment]
Rebased onto latest `master`.

**Comment by f1ames** [comment]
> > The other thing is that it uses dev integration version (I know that's how tests are run now via test:browser command), but I think it will be more reasonable to test production version/build. WDYT?
> 
> I think that's a very reasonable idea and also that we should add it as an option for the "normal" test run - we could then have two runs on bender (built version and dev version like for CKE4), so I'd rather extract it so a separate issue (this would also a good idea in other repos), WDYT?

Makes sense. This tester uses regular `npm run test` underneath so it is separated from testing procedure itself and the way it works. Since we have only command to run dev test it will be good to add another one to run tests on build version (which should be the default one I suppose).

As for having two runs dev/build, I would say for build version we should run "Ultimate tester" and for dev maybe latest version only to limit time needed for the CI runs? The most important is that production version works fine. And if it works, chances that dev one is broken are rather low.

File: .travis.yml
Comment (f1ames) [inline .travis.yml:23]: This triggers two separate jobs to be run on Travis CI (one with node 12 and the other with node 14). Do you think we really need this? :thinking:

File: scripts/react-tester.js
Comment (f1ames) [inline scripts/react-tester.js:31]: I wonder if we could go with better error logging? For example showing which versions failed. I know it can be deducted from browsing run logs but clear information on the log end would be useful too.

File: scripts/react-tester.js
Comment (f1ames) [inline scripts/react-tester.js:27]: Not sure if it's needed :thinking: Maybe assuming that if there is no error message means the run was successful will be enough?

**Review by f1ames — CHANGES_REQUESTED** [review]
Look good :+1: 

> I've limited versions tested according to this suggestion, so only the latest patch version for each minor release is tested. It cuts the number of builds from 33 to 15. Not sure if we need e.g. config option to enable also full run?

Makes sense :+1: However, the job still takes around 20 minutes which is quite long. OTOH I don't see many alternatives here (apart from running only latest major - which we support only one ATM, or selected by hand versions only). I think we could go with what I mentioned in https://github.com/ckeditor/ckeditor4-react/pull/166#issuecomment-800268575:

> As for having two runs dev/build, I would say for build version we should run "Ultimate tester" and for dev maybe latest version only to limit time needed for the CI runs? The most important is that production version works fine. And if it works, chances that dev one is broken are rather low.

The dev version would execute much faster (running only with latest supported React version) so we can get feedback quicker. And if it's green wait for multi-version tests.

File: .travis.yml
Comment (f1ames) [reply .travis.yml:23]: Hmm, and it seems v12 fails :thinking: Tried two times, both red, restarted 3rd time and we will see.

File: scripts/react-tester.js
Comment (Dumluregn) [reply scripts/react-tester.js:27]: I think it's nice to have the end message - you can be sure then that everything was executed, not just that there are no errors. This prevents the situation where for some reason the process was killed in the meantime and you might not notice it.

**Comment by Dumluregn** [comment]
>>> The other thing is that it uses dev integration version (I know that's how tests are run now via test:browser command), but I think it will be more reasonable to test production version/build. WDYT?

>> I think that's a very reasonable idea and also that we should add it as an option for the "normal" test run - we could then have two runs on bender (built version and dev version like for CKE4), so I'd rather extract it so a separate issue (this would also a good idea in other repos), WDYT?

> Makes sense. This tester uses regular npm run test underneath so it is separated from testing procedure itself and the way it works. Since we have only command to run dev test it will be good to add another one to run tests on build version (which should be the default one I suppose). @Dumluregn could you extract this issue.

Extracted here: https://github.com/ckeditor/ckeditor4-react/issues/184.

File: .travis.yml
Comment (Dumluregn) [reply .travis.yml:23]: I couldn't reproduce it using the same env as in failed tests (Safari `11.1.2`, Node `12`, React `16.3.2`). However it's unsure if in all 3 runs the errors were the same. I've pushed a commit with logging improvements so we'll see how it goes.
As for the very idea of checking both node versions - in a perfect world I think it would be totally justified. Unfortunately it 1) doubles the time needed for CI to finish checks (for now we can rarely run both builds simultaneously) and 2) seems so raise false alarms, so I'm not sure if we actually want it.

File: scripts/react-tester.js
Comment (Dumluregn) [reply scripts/react-tester.js]: Currently the process stops if any test fails - we are not moving forward in such case, so I've just added failing version number at the end of log (below the error so it should be easily visible). Do you think we should change this workflow?

I've also limited displaying the error information while maintaining the possibility of displaying the full info using `-v` flag.

**Comment by Dumluregn** [comment]
Now both Node versions failed and it seems it's the same test on the same browser: `does not raise element-conflict error` on Safari `11.1.2` (see [Node 12 log](https://travis-ci.com/github/ckeditor/ckeditor4-react/jobs/492027851#L370) and [Node 14 log](
https://travis-ci.com/github/ckeditor/ckeditor4-react/jobs/492027852#L1049)), so it looks we should take a look at it. BTW it would probably be a good idea to bump the Safari version on BS we use for testing 🤔  Maybe that would solve the issue.

**Comment by Dumluregn** [comment]
The build worked after the rerun but failed in the first cycle so I'm extracting an issue for the unstable test so that it didn't block us here as agreed with @f1ames. 
Therefore PR is ready for another review.

**Comment by Dumluregn** [comment]
PR is ready for review. Since @f1ames is off until the end of the week, @Comandeer if you find a while you could take it over.

File: scripts/react-tester.js
Comment (Comandeer) [inline scripts/react-tester.js:122]:
```suggestion
		execNpmCommand( 'install  --legacy-peer-deps', testPath );
		execNpmCommand( `install react@${ version } react-dom@${ version }  --legacy-peer-deps`, testPath );
```

This flag (`--legacy-peer-deps`) will make our tester compatible with npm 7+ (future proofing, FTW!)

**Review by Comandeer — COMMENTED** [review]
When I ran it locally, it failed with the following message:

```javascript
Successful tests:
[
  '16.0.0',  '16.1.1',
  '16.2.0',  '16.3.2',
  '16.4.2',  '16.5.2',
  '16.6.3',  '16.7.0',
  '16.8.6',  '16.9.0',
  '16.10.2', '16.11.0',
  '16.12.0', '16.13.1',
  '16.14.0'
]
Failed tests:
[]
```

🤔 

Overall it looks really good 👍🏻 Especially the new way of logging errors.

File: karma.conf.js
Comment (Dumluregn) [reply karma.conf.js:101]: I just wanted to check if it works and thought that the risk is lower with the older version 😛  I'll change it to the latest one.

**Comment by Dumluregn** [comment]
> When I ran it locally, it failed with the following message:

Yeah, I was naively checking if object is empty by comparing it to an empty object 🙈 :

```js
if ( errorLogs === {} ) {
```

I've fixed that, together with other comments. @Comandeer ready for rereview (or recommentation 🥠  <- this is meant to be a biscuit).

**Review by Comandeer — COMMENTED** [review]
LGTM! 🎉

**Comment by Dumluregn** [comment]
Since @Comandeer can't merge it as it is his own PR, I'm doing it.

**Pull request merged**
